### PR TITLE
lock voluntary exit domain

### LINKED ISSUE #7262: Lock voluntary exit domain
https://github.com/ethereum/consensus-specs/pull/3288

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description
Lock voluntary exit domain

## Fixed Issue(s)

fixes #7262 

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Review by StefanBratanov — COMMENTED** [review]
Hmm not sure that's correct. I think we should modify the `Fork` object passed to `verifyVoluntaryExitSignature` in `BlockProcessor.verifyVoluntaryExits` to be a hardcoded Capella one from Deneb onwards.

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/BlockProcessor.java [hunk 1]
```diff
@@ -168,6 +170,8 @@ void processWithdrawals(MutableBeaconState state, ExecutionPayloadSummary payloa
 
   Optional<List<Withdrawal>> getExpectedWithdrawals(BeaconState preState);
 
+  Bytes32 computeVoluntaryExitDomain(UInt64 epoch, Fork fork, Bytes32 getGenesisValidatorsRoot);
```
Comment (StefanBratanov) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/BlockProcessor.java]: I think I would change to passing the `BeaconState` and the `SignedVoluntaryExit` and do the getting the necessary stuff in the method

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/AbstractBlockProcessor.java [hunk 2]
```diff
@@ -838,14 +839,27 @@ protected BlockValidationResult verifyVoluntaryExits(
     for (SignedVoluntaryExit signedExit : exits) {
       boolean exitSignatureValid =
           operationSignatureVerifier.verifyVoluntaryExitSignature(
-              state.getFork(), state, signedExit, signatureVerifier);
+              state,
+              signedExit,
+              signatureVerifier,
+              computeVoluntaryExitDomain(
```
Comment (StefanBratanov) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/AbstractBlockProcessor.java]: think it's more readable to separate this in a variable inside the loop

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/operations/OperationSignatureVerifier.java [hunk 3]
```diff
@@ -107,9 +107,12 @@ public boolean verifyVoluntaryExitSignature(
       return false;
     }
 
+    /*
```
Comment (StefanBratanov) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/operations/OperationSignatureVerifier.java]: Can remove this I guess?

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/deneb/block/BlockProcessorDeneb.java [hunk 4]
```diff
@@ -98,6 +102,15 @@ public NewPayloadRequest computeNewPayloadRequest(final BeaconBlockBody beaconBl
     return new NewPayloadRequest(executionPayload, versionedHashes);
   }
 
+  @Override
+  public Bytes32 computeVoluntaryExitDomain(
+      UInt64 epoch, Fork fork, Bytes32 getGenesisValidatorsRoot) {
```
Comment (StefanBratanov) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/deneb/block/BlockProcessorDeneb.java]: just some finals 🗡️

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/BlockProcessor.java (hunk 1 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/BlockProcessor.java]: Done

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/AbstractBlockProcessor.java (hunk 2 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/AbstractBlockProcessor.java]: Done. In `Spec::verifyVoluntaryExitSignature` too

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/operations/OperationSignatureVerifier.java (hunk 3 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/operations/OperationSignatureVerifier.java]: My bad! Forgot it

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/deneb/block/BlockProcessorDeneb.java (hunk 4 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/deneb/block/BlockProcessorDeneb.java]: 🤺

**Review by StefanBratanov — APPROVED** [review]
LGTM from my side but reference tests seem to require a new release I think.

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/Spec.java [hunk 5]
```diff
@@ -461,9 +461,11 @@ public boolean verifyProposerSlashingSignature(
   public boolean verifyVoluntaryExitSignature(
       BeaconState state, SignedVoluntaryExit signedExit, BLSSignatureVerifier signatureVerifier) {
     final UInt64 epoch = signedExit.getMessage().getEpoch();
+    final Bytes32 voluntaryExitDomain =
+        getBlockProcessor(state.getSlot()).computeVoluntaryExitDomain(signedExit, state);
```
Comment (StefanBratanov) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/Spec.java]: Hmm actually I had a second look and think this should use the SpecVersion from the signedExit epoch so can do 

```java
final SpecVersion specVersion = atEpoch(epoch);
final Bytes32 domain =
        specVersion.getBlockProcessor().computeVoluntaryExitDomain(signedExit, state);
return specVersion.operationSignatureVerifier()....
```

Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/Spec.java]: Done

**Comment by mehdi-aouadi** [comment]
The new related reference tests updates will be released in a week or so. We could merge this PR when it's done

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/deneb/block/BlockProcessorDeneb.java [hunk 6]
```diff
@@ -98,6 +101,15 @@ public NewPayloadRequest computeNewPayloadRequest(final BeaconBlockBody beaconBl
     return new NewPayloadRequest(executionPayload, versionedHashes);
   }
 
+  @Override
+  public Bytes32 computeVoluntaryExitDomain(
+      final SignedVoluntaryExit signedVoluntaryExit, final BeaconState state) {
+    return miscHelpers.computeDomain(
+        Domain.VOLUNTARY_EXIT,
+        specConfig.toVersionCapella().orElseThrow().getCapellaForkVersion(),
```
Comment (StefanBratanov) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/deneb/block/BlockProcessorDeneb.java]: Just a suggestion in meantime to change this to `SpecConfigCapella.required(specConfig).getCapellaForkVersion()` (clearer error message)

Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/deneb/block/BlockProcessorDeneb.java]: Yep better!

**Pull request merged**